openpype/hosts/maya/api/lib.py

```diff
@@ -3129,31 +3137,53 @@ def remove_render_layer_observer():
         pass
 
 
-def update_content_on_context_change():
+def set_attribute_instances(attribute, value):
+    """Set attribute value of publishable instances (their objectSets)
     """
-    This will update scene content to match new asset on context change
+    collected_instances = cmds.ls(
+        "*.id",
+        long=True,
+        type="objectSet",
+        recursive=True,
+        objectsOnly=True
+    )
+
+    for instance in collected_instances:
+        id_attr = "{}.id".format(instance)
+        if cmds.getAttr(id_attr) != "pyblish.avalon.instance":
+            continue
+
+        if not cmds.attributeQuery(attribute, node=instance, exists=True):
+            continue
+
+        attr = "{}.{}".format(instance, attribute)
+
+        if cmds.getAttr(attr, type=True) == "string":
+            cmds.setAttr(attr, value, type="string")
+        else:
+            cmds.setAttr(attr, value)
+
+
+def update_instances_asset_attribute():
+    """Update 'asset' attribute of publishable instances (their objectSets)
+    that got one.
+    """
+
+    set_attribute_instances("asset", get_current_project_asset()['name'])
+
+
+def update_instances_frame_range():
+    """Update 'frameStart', 'frameEnd', 'handleStart', 'handleEnd' and 'fps'
+    attributes of publishable instances (their objectSets) that got one.
     """
-    scene_sets = cmds.listSets(allSets=True)
     asset_doc = get_current_project_asset()
-    new_asset = asset_doc["name"]
-    new_data = asset_doc["data"]
-    for s in scene_sets:
-        try:
-            if cmds.getAttr("{}.id".format(s)) == "pyblish.avalon.instance":
-                attr = cmds.listAttr(s)
-                print(s)
-                if "asset" in attr:
-                    print("  - setting asset to: [ {} ]".format(new_asset))
-                    cmds.setAttr("{}.asset".format(s),
-                                 new_asset, type="string")
-                if "frameStart" in attr:
-                    cmds.setAttr("{}.frameStart".format(s),
-                                 new_data["frameStart"])
-                if "frameEnd" in attr:
-                    cmds.setAttr("{}.frameEnd".format(s),
-                                 new_data["frameEnd"],)
-        except ValueError:
-            pass
+    asset_data = asset_doc["data"]
+
+    set_attribute_instances("frameStart", asset_data["frameStart"])
+    set_attribute_instances("frameEnd", asset_data["frameEnd"])
+    set_attribute_instances("handleStart", asset_data["handleStart"])
+    set_attribute_instances("handleEnd", asset_data["handleEnd"])
+    set_attribute_instances("fps", asset_data['fps'])
```

Comment:
This will be very slow - we're querying all the instances every time. This is far from the recommended approach.

Here's a more optimal approach.

This only queries the asset if we need it.
This only queries the related data.
This only tries to find the instances once.
```python
def iter_publish_instances():
    for node in cmds.ls(
        "*.id",
        long=True,
        type="objectSet",
        recursive=True,
        objectsOnly=True
    )
        id_attr = "{}.id".format(instance)
        if cmds.getAttr(id_attr) != "pyblish.avalon.instance":
            continue

       yield node

def update_instances_frame_range():
    attributes = ["frameStart", "frameEnd", "handleStart", "handleEnd", "fps"]
    attrs_per_instance = {}
    for instance in iter_publish_instances():
        instance_attrs = []
        for attr in attributes:
            if cmds.attributeQuery(attr, node=node, exists=True):
                instance_attrs .append(attr)
        if instance_attrs :
            attrs_per_instance [instance] = instance_attrs

    if not attrs_per_instance:
        # no instances with any frame related attributes
        return

    fields = ["data.{}".format(key) for key in attributes]
    asset_doc = get_current_project_asset(fields=fields)
    asset_data = asset_doc["data"]
    for node, attrs in attrs_per_instance.items():
        for attr in attrs:
            plug = "{}.{}".format(node, attr)
            value = asset_data[attr]
            if cmds.getAttr(plug, type=True) == "string":
                cmds.setAttr(plug, value, type="string")
            else:
                cmds.setAttr(plug, value)
```